### PR TITLE
Like replies

### PR DESCRIPTION
**Pull request by BobbyBorisov**
This is a WIP implementation for #334

### CODE REVIEW

**Comment by BobbyBorisov** [comment]
@driesvints could you take a look when you are free :)

**Comment by driesvints** [comment]
@BobbyBorisov will do. Btw, you can always request a review from me by using the "reviewers" menu in the right sidebar ----->

In the meantime, can you already fix the StyleCI issues? Thanks!

**Comment by BobbyBorisov** [comment]
@driesvints Sure, fixing them right now. Btw, option for selecting reviewers is disabled for me. I need some permissions probably? ;)

**Comment by driesvints** [comment]
@BobbyBorisov you're right! Didn't realize it was limited to write access.

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,5 @@ Homestead.yaml
 .env
 /aliases
 /after.sh
+.idea/
```
Comment (BobbyBorisov) [inline .gitignore]: maybe its better to use global gitignore file?

**Comment by BobbyBorisov** [comment]
Hey @driesvints Happy New Year First! Wish you great talks like Laracon EU 2017 and many apps with less bugs! 

I wonder would you have some free time to review this? Currently what remains is connecting the backend with frontend(I made like/unlike buttons vue components).

Cheers!

File: resources/assets/js/components/Like.vue [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        <span style="margin-right:3px">{{count}}</span><i class="fa" :class="classes" @click="toggle"></i>
+    </div>
+</template>
+
+<script>
+    export default {
+        data(){
+            return {
+                count:0,
+                isLiked: false
```
Comment (m1guelpf) [inline resources/assets/js/components/Like.vue]: How about getting the initial status using a prop?

File: resources/views/forum/threads/show.blade.php [hunk 3]
```diff
@@ -71,6 +71,11 @@
                                 <a class="btn btn-danger btn-xs" href="#" data-toggle="modal" data-target="#deleteReply{{ $reply->id() }}">Delete</a>
                             </div>
                         @endcan
+                        @auth
+                            <div class="thread-info-likes">
+                                <like></like>
```
Comment (m1guelpf) [inline resources/views/forum/threads/show.blade.php]: How do we know which thread we're at?

File: tests/Components/Jobs/DislikeReplyTest.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        $reply = factory(Reply::class)->create();
+
+        $reply->likedBy($user);
+        $this->assertTrue($reply->fresh()->isLikedBy($user));
+
+        $this->dispatch(new DislikeReply($reply, $user));
+
+        $this->dispatch(new DislikeReply($reply, $user));
+
+        //$this->assertFalse($reply->fresh()->isLikedBy($user));
```
Comment (m1guelpf) [inline tests/Components/Jobs/DislikeReplyTest.php]: Why is this commented out?

File: resources/assets/js/components/Like.vue (hunk 2 above)
Comment (BobbyBorisov) [reply resources/assets/js/components/Like.vue]: Yes, i agree, i haven't fully implemented like vue component, will do today or tomorrow

File: resources/views/forum/threads/show.blade.php (hunk 3 above)
Comment (BobbyBorisov) [reply resources/views/forum/threads/show.blade.php]: not fully implemented, will do ;)

File: app/Helpers/HasLikes.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        if ($user == null) {
+            return false;
+        }
+
+        return $this->likes()->where('user_id', $user->id)->exists();
+    }
+
+    public function getLikesCountAttribute()
+    {
+        return $this->likes->count();
```
Comment (deleugpn) [inline app/Helpers/HasLikes.php]: `$this->likes()->count()` would also work and avoid lazy-loading?

File: app/Helpers/HasLikes.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    protected static function bootHasLikes()
+    {
+        static::deleting(function ($model) {
+            $model->likes->each->delete();
+        });
+    }
+
+    public function likedBy(User $user)
+    {
+        $this->likes()->create(['user_id' => $user->id]);
```
Comment (deleugpn) [inline app/Helpers/HasLikes.php]: Perhaps would be nice to define an abstract method for `likes()`?

**Comment by driesvints** [comment]
@BobbyBorisov heya. I'm really sorry that I didn't respond sooner. I'm about to launch thread subscriptions and plan on implementing most of the PR's, which are currently open, after that. Thanks for your patience! 🙏

**Comment by joedixon** [comment]
This is still WIP, but wanted to show progress so far.
Updated likes to be a similar style to Github reactions and made them available on both the thread and the replies.
Rather than make this a Vue component, is standard blade with a page refresh for the time being.

Screenshot 2019-11-24 at 21 27 12

**Comment by driesvints** [comment]
@joedixon looks good!

**Comment by joedixon** [comment]
Hey @driesvints - this is good for review

**Comment by driesvints** [comment]
@joedixon thanks. I'll try to review this on saturday.

File: app/Helpers/ReceivesReplies.php [hunk 7]
```diff
@@ -25,7 +25,11 @@ public function latestReplies(int $amount = 5)
 
     public function deleteReplies()
     {
-        $this->repliesRelation()->delete();
+        // We need to explicitly iterate over the repies and delete them
+        // separately because all related models need to be deleted.
+        foreach ($this->repliesRelation()->get() as $reply) {
+            $reply->delete();
+        }
```
Comment (driesvints) [inline app/Helpers/ReceivesReplies.php]: Bit unfortunate that we can't use foreign keys for this. Beginning to dislike polymorphic relationships 😄

**Comment by driesvints** [comment]
Merged and deployed. Thanks @BobbyBorisov and @joedixon!

**Pull request merged**